transformers4rec/torch/model/prediction_task.py

```diff
@@ -417,50 +438,249 @@ class _NextItemPredictionTask(torch.nn.Module):
+        That probability can be approximated by by 1 - (1 - p)^n
+        and we use a numerically stable version: -expm1(num_tries * log1p(-p))
+        """
+        return (-(-dist.double().log1p_() * n_sample).expm1_()).float()
+
+    def sample(self, labels: torch.Tensor):
+        """Sample negative samples and calculate their probabilities.
+
+        If `unique_sampling==True`, then only unique sampled items will be returned.
+        The actual # samples will vary from run to run if `unique_sampling==True`,
+        as sampling without replacement (`torch.multinomial(..., replacement=False)`) is slow,
+        so we use `torch.multinomial(..., replacement=True).unique()`
+        which doesn't guarantee the same number of unique sampled items.
+        You can try to increase n_samples_multiplier_before_unique
+        to increase the chances to have more unique samples in that case.
+
+        Parameters
+        ----------
+        labels : torch.Tensor, dtype=torch.long, shape=(batch_size,)
+            The input labels for which negative samples should be generated.
+
+        Returns
+        -------
+        neg_samples : torch.Tensor, dtype=torch.long, shape=(n_samples,)
+            The unique negative samples drawn from the log-uniform distribution.
+        true_probs : torch.Tensor, dtype=torch.float32, shape=(batch_size,)
+            The probabilities of the input labels according
+            to the log-uniform distribution (depends on self.unique_sampling choice).
+        samp_log_probs : torch.Tensor, dtype=torch.float32, shape=(n_samples,)
+            The probabilities of the sampled negatives according
+            to the log-uniform distribution (depends on self.unique_sampling choice).
+        """
+
+        if not torch.is_tensor(labels):
+            raise TypeError("Labels must be a torch.Tensor.")
+        if labels.dtype != torch.long:
+            raise ValueError("Labels must be a tensor of dtype long.")
+        if labels.dim() > 2 or (labels.dim() == 2 and min(labels.shape) > 1):
+            raise ValueError(
+                "Labels must be a 1-dimensional tensor or a 2-dimensional tensor"
+                "with one of the dimensions equal to 1."
+            )
+        if labels.size(0) == 0:
+            raise ValueError("Labels must not be an empty tensor.")
+        if (labels < 0).any() or (labels > self.max_id).any():
+            raise ValueError("All label values must be within the range [0, max_id].")
+
+        n_tries = self.n_sample
+
+        with torch.no_grad():
+            neg_samples = torch.multinomial(self.dist, n_tries, replacement=True).unique()[
+                : self.max_n_samples
+            ]
+
+            device = labels.device
+            neg_samples = neg_samples.to(device)
+
+            if self.unique_sampling:
+                dist = self.unique_sampling_dist
+            else:
+                dist = self.dist
+            dist = dist.to(device)
```

Comment:
Would it be possible to move the definition of dist to the class constructor, to avoid copying the tensor to the GPU/CPU device multiple times?

Reply:
The challenge is how to get the device in the constructor. Any ideas?